File: leet100/01_hash/leet_02_049.py
class Solution(object):
    def groupAnagrams(self, strs):
        """
        :type strs: List[str]
        :rtype: List[List[str]]
        """
        orig2strs = {}
        for str in strs:
            ss = ''.join(sorted(str))
            if ss not in orig2strs:
                orig2strs[ss] = []
            orig2strs[ss].append(str)

        results = []
        # for value in orig2strs.values():
        #     results.append(value)
        for key, value in orig2strs.items():
            results.append(value)

        return results

if '__name__ == __main__':
    s = Solution()
    strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
    results = s.groupAnagrams(strs)
    print(results)

File: leet100/01_hash/leet_02_049.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <map>

#include "../../cpp/utils/str_tools.h"


class Solution {
public:
    std::vector<std::vector<std::string>> groupAnagrams(std::vector<std::string>& strs) 
    {
        std::map<std::string, std::vector<std::string>> sorted_2_strs;
        for (auto& str : strs) {
            auto str_copy = str;
            std::sort(str_copy.begin(), str_copy.end());
            sorted_2_strs[str_copy].push_back(str);
        }

        std::vector<std::vector<std::string>> results;
        for (auto& item : sorted_2_strs) {
            results.push_back(item.second);
        }
        return results;
    }
};

void test1()
{
    Solution sn;
    std::vector<std::string> strs {"eat", "tea", "tan", "ate", "nat", "bat"};
    std::vector<std::vector<std::string>> results = sn.groupAnagrams(strs);
    std::cout << "[['bat'],['nat','tan'],['ate','eat','tea']]" << std::endl;
    for (auto& vec : results) {
        std::cout << StringTools::ToString(vec) << std::endl;
    }
}

int main()
{
    test1();
    return 0;
}